cpmpy/transformations/decompose_global.py

```diff
@@ -1,107 +1,134 @@
 import copy
 
-from ..expressions.globalconstraints import GlobalConstraint
-from ..expressions.core import Expression, Comparison, Operator
+from .normalize import toplevel_list
+from ..expressions.globalconstraints import GlobalConstraint, DirectConstraint
+from ..expressions.core import Expression, Comparison, Operator, BoolVal
+from ..expressions.variables import _BoolVarImpl, intvar, boolvar, _NumVarImpl, cpm_array, NDVarArray
 from ..expressions.utils import is_any_list, eval_comparison
 from ..expressions.python_builtins import all
-from .flatten_model import flatten_constraint
 
-def decompose_global(lst_of_expr, supported=set(), supported_reif=set()):
 
+def decompose_in_tree(lst_of_expr, supported=set(), supported_nested=set(), _toplevel=None, nested=False):
     """
         Decomposes any global constraint not supported by the solver
-        Accepts a list of flat constraints as input
-        Returns a list of flat constraints
+        Accepts a list of CPMpy expressions as input and returns a list of CPMpy expressions,
+            otherwise, returns a list of CPMpy expressions and new constraints to be added toplevel.
 
         - supported: a set of supported global constraints or global functions
-        - supported_reified: a set of supported global constraints within a reification
+        - supported_nested: as set of supported global constraints/functions within other expressions
```

Comment:
I don't understand the need for this. All expressions in CPMpy can be used 'nested' (either reified or as part of a numeric operation)? why does this need special handling?

Reply:
Solvers can support differnet global constraint depending on the context. I.e., AllDiff is supported by OR-tools toplevel but not in a nested (reified) context. Therefore, we need the distinction.